maven-core/src/main/java/org/apache/maven/lifecycle/internal/LifecycleDependencyResolver.java

```diff
@@ -90,7 +90,13 @@ public LifecycleDependencyResolver(ProjectDependenciesResolver projectDependenci
 
     public static List<MavenProject> getProjects(MavenProject project, MavenSession session, boolean aggregator) {
         if (aggregator && project.getCollectedProjects() != null) {
-            return getProjectAndSubModules(project).collect(Collectors.toList());
+            List<MavenProject> downstreamProjects = session.getProjectDependencyGraph()
+                    .getDownstreamProjects(project, true); // sorted but more than needed
+            List<MavenProject> projectAndSubmodules =
+                    getProjectAndSubModules(project).collect(Collectors.toList()); // not sorted but what we need
+            return downstreamProjects.stream()
+                    .filter(projectAndSubmodules::contains)
+                    .collect(Collectors.toList()); // sorted what we need
```

Comment:
The returned list should contain the `project` instance.

Reply:
fixed

Reply:
I can propose the following which may be slightly easier to read
```
            List<MavenProject> projectAndSubmodules =
                    getProjectAndSubModules(project).collect(Collectors.toList());
            return Stream.concat(
                            Stream.of(project),
                            session.getProjectDependencyGraph().getDownstreamProjects(project, true).stream())
                    .filter(projectAndSubmodules::contains)
                    .collect(Collectors.toList());
```

Reply:
This is exactly what I figured out yesterday as well